### ISSUE TITLE
`fury build compile` hangs without CPU usage

### ISSUE THREAD

**Issue by ArturGajowy** [propensive/fury-old#229]
I was following the "Getting Started" tutorial, and after not being able to receive the compilation errors after 30 s, I've noticed my CPU's is unused.

I've then tried building the `propensive/.shapeless` layer, with same result (see log below).

This must be some setup issue. How do I get more output from fury / go through the logs?

What software should I get rid of to make sure I have a clean fury install?

```
y510p% git clone git@github.com:propensive/.shapeless.git
Cloning into '.shapeless'...
remote: Enumerating objects: 29, done.
remote: Counting objects: 100% (29/29), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 29 (delta 7), reused 28 (delta 6), pack-reused 0
Receiving objects: 100% (29/29), 12.68 KiB | 0 bytes/s, done.
Resolving deltas: 100% (7/7), done.
Checking connectivity... done.

y510p% cd .shapeless 

y510p% time fury
Cloning Git repository git@github.com:propensive/.scala.git.
Checking out all sources.
Cloning Git repository git@github.com:milessabin/shapeless.git.
Checking out [core/src/main/scala, core/src/main/scala_2.11+/shapeless, core/src/main/scala_2.13-/shapeless].
^C shapeless/gen      .  [ ]
fury  0,13s user 0,04s system 0% cpu 57,305 total
y510p% 
```

**Comment by ArturGajowy** [comment]
I've just done a clean (?) install of fury-0.4.0 after deleting `~/.bloop`, and killing all processes containing `bloop` or `nail` in their path. The Fury installer re-installed bloop for me. The result is same :(

**Comment by tpasternak** [comment]
It's still a known issue, current workaround is to kill bloop and blp-server.

**Comment by ArturGajowy** [comment]
@tpasternak Thx! Can you point me to the original description of the issue and the workaround?

I've killed `blp-server`, but then fury says:

```
artur@y510p:~/projects/.shapeless$ fury

Starting bloop compile serverAn unexpected error occurred which has been logged to /home/artur/projects/.shapeless/.fury/logs/4c70b1b9-7c40-4279-8f02-8c12c1e92d94.log.
```

The logs never came, though (!)

```
artur@y510p:~/projects/.shapeless$ cat /home/artur/projects/.shapeless/.fury/logs/4c70b1b9-7c40-4279-8f02-8c12c1e92d94.log
cat: /home/artur/projects/.shapeless/.fury/logs/4c70b1b9-7c40-4279-8f02-8c12c1e92d94.log: No such file or directory
```

**Comment by tpasternak** [comment]
You're right, it was not submitted yet, thanks :)

Btw are you using fury 0.4.0?

**Comment by ArturGajowy** [comment]
Yup :)

**Comment by tpasternak** [comment]
Ok, I managed to achieve the same effect by killing all bloop processes and removing ~/.bloop directory. Could you check if bloop is now in your path? (Definitely we have to fix it but for now we're looking for the clear reproduction scenario)

**Comment by ArturGajowy** [comment]
I can provide a repro scenario in a docker container very easily. Bloop is not in my path, Fury installed it for me. I'd assume it'll know where it is then ;) I'll provide the repro in a minute

**Comment by ArturGajowy** [comment]
Here it goes: https://gist.github.com/ArturGajowy/92b132760a71c5fdaf22e655ec6cdbbe

TL;DR:
- `docker run -it openjdk:11-jdk bash` opens a shell that has all that is needed to install fury
- there's no `source` command, so `. ~/.bashrc` after installing Fury
- I `git clone https://github.com/ArturGajowy/.shapeless.git` because fury uses SSH remote refs and I've changed them to https to be able to checkout without having a private key that's being actively used on github in that docker image (I'll file an issue for that later, as it prohibits/hinders usage in CI)

**Comment by ArturGajowy** [comment]
Adding `export PATH="$HOME/.bloop/:$PATH"` to `.bashrc` and figuring out a sequence for killing bloop instances did the trick for me, although I ended up with compilation errors for the `.shapeless` layer from my fork.

**Comment by ArturGajowy** [comment]
I could reproduce the compilation errors in the docker container as well as on my local machine.

The sequence for killing bloop seems to be:
kill all of: bloop, blp-server, fury
run fury to see 'can't connect to bloop' message and a non-zero exit
run fury (starts compiling)

**Comment by tpasternak** [comment]
To sum up: we have a few problems:
1. The PATH is not set in the installation script (if we set PATH for `fury` we should consequently set it for `bloop`
2. When `bloop` is already started, `fury` hangs

Btw in this Dockerfile your shapeless clone is built without errors https://gist.github.com/tpasternak/749a6791bcb5d30331cfd28cb223bb1a

Thank you for the detailed descriptions and examples!

**Comment by ArturGajowy** [comment]
I'm not sure about the conclusion in 1. - maybe Fury should be referencing `bloop` directly instead of relying on PATH? While not relying on PATH seems more robust - both approaches seem fine to me.

Thx for the docker file! Did you manage to reproduce the compilation issues I mentioned either locally or by pasting commands into interactive bash session on the docker image?

I'm curious what difference there is that makes the `.shapeless` layer compile when run during a `docker build`...

**Comment by tpasternak** [comment]
> I'm not sure about the conclusion in 1. - maybe Fury should be referencing `bloop` directly instead of relying on PATH? While not relying on PATH seems more robust - both approaches seem fine to me.

It's not clear for me what should be the _right_ solution. The bloop's `launcher` checks whether bloop is in the `PATH` or `~/.bloop` and if it is not, it installs it to `~/.bloop`. 

> Thx for the docker file! Did you manage to reproduce the compilation issues I mentioned either 
locally or by pasting commands into interactive bash session on the docker image?
>
> I'm curious what difference there is that makes the `.shapeless` layer compile when run during a `docker build`...

I just tried it - yes, I reproduced the issues. It seems that the workaround `kill` bloop and `fury restart` after entering `.shapeless` dir. So we have three issues on the table. The difference probably comes from new bash sessions starts in the Dockerfile

**Comment by tpasternak** [comment]
https://gist.github.com/tpasternak/6eb8524b4971ddbb9e337cc30391225e

This dockerfile (with line 10. removed) shows the last issue

**Comment by tpasternak** [comment]
Final docker file to reproduce hanging: https://gist.github.com/tpasternak/7a839eb9309a2dfce74e06f0cbec130b

Steps:
1. `fury start` **without** `bloop` in path
2. Start bloop server
3. Run `fury`
